src/getValidators.ts

```diff
@@ -0,0 +1,179 @@
+import {
+  Address,
+  Chain,
+  Hex,
+  PublicClient,
+  Transport,
+  decodeFunctionData,
+  getAbiItem,
+  getFunctionSelector,
+} from 'viem';
+import { rollupCreator, upgradeExecutor } from './contracts';
+import { rollupAdminLogicABI, safeL2ABI } from './abi';
+import { createRollupFetchTransactionHash } from './createRollupFetchTransactionHash';
+
+const createRollupABI = getAbiItem({ abi: rollupCreator.abi, name: 'createRollup' });
+const createRollupFunctionSelector = getFunctionSelector(createRollupABI);
+
+const setValidatorABI = getAbiItem({ abi: rollupAdminLogicABI, name: 'setValidator' });
+const setValidatorFunctionSelector = getFunctionSelector(setValidatorABI);
+
+const executeCallABI = getAbiItem({ abi: upgradeExecutor.abi, name: 'executeCall' });
+const upgradeExecutorExecuteCallFunctionSelector = getFunctionSelector(executeCallABI);
+
+const execTransactionABI = getAbiItem({ abi: safeL2ABI, name: 'execTransaction' });
+const safeL2FunctionSelector = getFunctionSelector(execTransactionABI);
+
+const ownerFunctionCalledEventAbi = getAbiItem({
+  abi: rollupAdminLogicABI,
+  name: 'OwnerFunctionCalled',
+});
+
+function getValidatorsFromFunctionData<
+  TAbi extends (typeof createRollupABI)[] | (typeof setValidatorABI)[],
+>({ abi, data }: { abi: TAbi; data: Hex }) {
+  const { args } = decodeFunctionData({
+    abi,
+    data,
+  });
+  if (typeof args === 'undefined') {
+    return [];
+  }
```

Comment:
looks like `args` will never be `undefined` after you tightened up the type so this can be deleted

Reply:
Good catch